packages/teleterm/src/ui/DocumentCluster/ClusterResources/useServerSideResources.ts

```diff
@@ -0,0 +1,149 @@
+import { useState, useEffect } from 'react';
+import { SortType } from 'design/DataTable/types';
+import { useAsync } from 'shared/hooks/useAsync';
+import { ServerSideParams } from 'teleterm/services/tshd/types';
+import { useAppContext } from 'teleterm/ui/appContextProvider';
+import { retryWithRelogin } from 'teleterm/ui/utils';
+import { useClusterContext } from '../clusterContext';
+import { AgentFilter, AgentLabel } from 'teleport/services/agents';
+import { addAgentLabelToQuery } from 'shared/utils/addAgentLabelToQuery';
+import { ResourceKind } from 'teleport/Discover/Shared';
+
+export function useServerSideResources<Agent>(
+  fetchFunction: (params: ServerSideParams) => Promise<{
+    agentsList: Array<Agent>;
+    totalCount: number;
+    startKey: string;
+  }>
+) {
+  const ctx = useAppContext();
+  const { clusterUri, documentUri } = useClusterContext();
+  const [pageIndex, setPageIndex] = useState(0);
+  const [keys, setKeys] = useState<string[]>([]);
+  const [agentFilter, setAgentFilter] = useState<AgentFilter>({
+    sort: getDefaultSort(),
+  });
+  const [pageCount, setPageCount] = useState({
+    to: 0,
+    from: 0,
+    total: 0,
+  });
+
+  const limit = 15;
+
+  // startKey is used here as a way to paginate through agents returned from
+  // their respective rpcs.
+  const [fetchAttempt, fetch] = useAsync(async (startKey: string) =>
+    retryWithRelogin(ctx, documentUri, clusterUri, () =>
+      fetchFunction({
+        ...agentFilter,
+        limit,
+        clusterUri,
+        startKey,
+      })
+    )
+  );
+
+  // If there is no startKey at the current page's index, there is no more data to get
+  function nextPage() {
+    const proposedKey = keys[pageIndex];
+    if (proposedKey) {
+      return () => {
+        fetch(proposedKey);
+        setPageIndex(pageIndex + 1);
+      };
+    }
+    return null;
+  }
+
+  // If we are at the first page (index 0), we cannot fetch more previous data
+  function prevPage() {
+    const newPageIndex = pageIndex - 1;
+    if (newPageIndex > -1) {
+      return () => {
+        // if we got pageIndex - 1, we'd get the same data as the previous "next" we just called.
+        // we have to go back 2. which is why we use newPageIndex instead
+
+        // example after going from pageIndex = 0 to pageIndex = 1:
+        // ['startKeyFromInitialFetch', 'goEvenFurtherStartKey']
+        //     ^- just came from here          ^- we are here
+        //
+        // so in order to get the start key for the page BEFORE this one, we go back twice.
+        // this could also be written/read as keys[pageIndex - 2]
+        fetch(keys[newPageIndex - 1]);
+        setPageIndex(newPageIndex);
```

Comment:
The new implementation is better than the one from bd7fc4bcd28b. However, I remember we talked about keeping the implementation of `prevPage` and `nextPage` as lean as possible where they could merely update the pagination metadata, which would cause some `useEffect` to re-run which would actually fetch the data. Did that turn out to not be a good approach here?

Reply:
I think my brain got a bit stuck in an old flow and when I finally figured out how to get it 'lean' it stayed with the old. Updated the `fetch` useEffect to react to agentFilter and page. (page will always update with agentFilter after your feedback above, but agentFilter won't always update with page).